crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -143,49 +205,440 @@ impl Acir {
+                }
+            },
+            Operation::Load(array_idx) => {
+                //retrieves the value from the map if address is known at compile time:
+                //address = l_c and should be constant
+                if let Some(val) = l_c.to_const() {
+                    let address = mem::Memory::as_u32(val);
+                    if self.memory_map.contains_key(&address) {
+                        InternalVar::from(&self.memory_map[&address].expression.clone())
+                    } else {
+                        //if not found, then it must be a witness (else it is non-initialised memory)
+                        let array = &cfg.mem.arrays[array_idx as usize];
+                        let index = (address - array.adr) as usize;
+                        let w = if array.witness.len() > index {
+                            array.witness[index]
+                        } else {
+                            self.memory_witness[&array_idx][index]
+                        };
+                        InternalVar::from(&w)
+                    }
+                } else {
+                    todo!("dynamic arrays are not implemented yet");
+                }
             }
-        };
 
-        let output_var = InternalVar {
-            expression: output,
-            //value: FieldElement::from(0_u32),
-            idx: Some(ins.idx),
-            witness: None, //TODO put the witness when it exist
+            Operation::Store(_) => {
+                //maps the address to the rhs if address is known at compile time
+                if let Some(val) = r_c.to_const() {
+                    let address = mem::Memory::as_u32(val);
+                    self.memory_map.insert(address, l_c);
+                    dbg!(&self.memory_map);
+                    //we do not generate constraint, so no output.
+                    InternalVar::default()
+                } else {
+                    todo!("dynamic arrays are not implemented yet");
+                }
+            }
         };
 
+        output_var.idx = Some(ins.idx);
+
         self.arith_cache.insert(ins.idx, output_var);
     }
+
+    pub fn print_field(f: FieldElement) -> String {
+        if f == -FieldElement::one() {
+            return "-".to_string();
+        }
+        if f == FieldElement::one() {
+            return String::new();
+        }
+        let big_f = BigUint::from_bytes_be(&f.to_bytes());
+
+        if big_f.to_string() == "4294967296" {
+            return "2^32".to_string();
+        }
+        let s = big_f.bits();
+        let big_s = BigUint::one() << s;
+        if big_s == big_f {
+            return format!("2^{}", s);
+        }
+        if big_f == BigUint::zero() {
+            return "0".to_string();
+        }
+        if big_f.clone() % BigUint::from(2_u128).pow(32) == BigUint::zero() {
+            return format!("2^32*{}", big_f / BigUint::from(2_u128).pow(32));
```

Comment:
Because of bit sizes, there are a lot of powers of 2. The overflow strategy is adding many powers of 2 up to the field characteristic. While we can easily spot powers of 2 until 2^16, it is becoming harder after that.